src/Type/Php/StrCaseFunctionsReturnTypeExtension.php

```diff
@@ -0,0 +1,82 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\ShouldNotHappenException;
+use PHPStan\Type\Accessory\AccessoryLiteralStringType;
+use PHPStan\Type\Accessory\AccessoryNonEmptyStringType;
+use PHPStan\Type\Accessory\AccessoryNumericStringType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\IntersectionType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use function count;
+use function in_array;
+use function is_callable;
+
+class StrCaseFunctionsReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return in_array($functionReflection->getName(), [
+			'strtoupper',
+			'strtolower',
+			'mb_strtoupper',
+			'mb_strtolower',
+			'lcfirst',
+			'ucfirst',
+			'ucwords',
+		], true);
+	}
+
+	public function getTypeFromFunctionCall(
+		FunctionReflection $functionReflection,
+		FuncCall $functionCall,
+		Scope $scope,
+	): Type
+	{
+		$args = $functionCall->getArgs();
+		if (count($args) < 1) {
+			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		}
+
+		$argType = $scope->getType($args[0]->value);
+
+		if ($argType instanceof ConstantStringType) {
```

Comment:
You need `TypeUtils::getConstantStrings()` to account for unions.

Reply:
after I got the repro for bug6987 running it got obvious  ;-)